File: PlayerController.js
/**
 * PlayerController.js - Handles all player movement, input, and ball interaction
 * Features:
 * - 8-directional movement with smooth acceleration/deceleration
 * - Momentum-based physics for realistic feel
 * - Dribbling mechanics (ball stays near feet with physics separation)
 * - Sprint capability with stamina drain
 * - Passing and shooting input preparation
 */

class PlayerController {
    constructor(scene, position, playerTeam, playerNumber, isLocalPlayer = false) {
        // ===== References =====
        this.scene = scene;
        this.ball = null; // Will be set by game manager
        
        // ===== Transform Data =====
        this.position = new THREE.Vector3(position.x, position.y, position.z);
        this.velocity = new THREE.Vector3(0, 0, 0);
        this.currentVelocity = new THREE.Vector3(0, 0, 0);
        this.rotation = 0; // Rotation around Y axis in radians
        
        // ===== Movement Parameters =====
        this.maxSpeed = 12;              // Maximum walking speed (m/s)
        this.maxSprintSpeed = 18;        // Maximum sprint speed (m/s)
        this.acceleration = 20;          // How quickly player reaches max speed
        this.deceleration = 15;          // How quickly player slows down
        this.rotationSpeed = 10;         // Smooth rotation towards movement direction
        this.dribbleSlowDownFactor = 0.85; // Speed reduction when dribbling
        
        // ===== Dribbling Parameters =====
        this.dribbleControlRadius = 1.5;    // How far ball can drift from feet
        this.dribbleDragStrength = 8;       // How strongly player pulls ball back
        this.dribbleMaxBallDistance = 2.5;  // Max distance before losing possession
        
        // ===== Stamina System =====
        this.maxStamina = 100;
        this.currentStamina = 100;
        this.staminaDrainSprint = 30;  // Stamina drain per second while sprinting
        this.staminaRecovery = 15;     // Stamina recovery per second
        this.canSprint = true;
        
        // ===== Input Handling =====
        this.inputDirection = new THREE.Vector2(0, 0);
        this.isSprinting = false;
        this.isDribbling = false;
        
        // ===== Player State =====
        this.playerTeam = playerTeam;
        this.isLocalPlayer = isLocalPlayer;
        this.playerNumber = playerNumber;
        this.hasBallPossession = false;
        this.lastTouchTime = 0;
        
        // ===== 3D Mesh =====
        this.createPlayerMesh();
        
        // ===== Input Keys Tracking =====
        this.keysPressed = {
            W: false,
            A: false,
            S: false,
            D: false,
            Shift: false,
            Space: false,
            Control: false
        };
    }

    /**
     * Create 3D player mesh
     */
    createPlayerMesh() {
        // Create player capsule (simple cylinder + sphere for head)
        const bodyGeometry = new THREE.CylinderGeometry(0.3, 0.3, 1.5, 16);
        const headGeometry = new THREE.SphereGeometry(0.2, 16, 16);
        
        // Team color (Red for Team A, Blue for Team B)
        const teamColor = this.playerTeam === gameStateManager.Team.TeamA ? 0xFF3333 : 0x3366FF;
        const material = new THREE.MeshPhongMaterial({ color: teamColor });
        
        // Create mesh group
        this.mesh = new THREE.Group();
        
        // Body
        const body = new THREE.Mesh(bodyGeometry, material);
        body.position.y = 0.75;
        this.mesh.add(body);
        
        // Head
        const head = new THREE.Mesh(headGeometry, material);
        head.position.y = 1.7;
        this.mesh.add(head);
        
        // Jersey number label
        const canvas = document.createElement('canvas');
        canvas.width = 64;
        canvas.height = 64;
        const ctx = canvas.getContext('2d');
        ctx.fillStyle = '#FFFFFF';
        ctx.font = 'bold 48px Arial';
        ctx.textAlign = 'center';
        ctx.fillText(this.playerNumber.toString(), 32, 48);
        
        const texture = new THREE.CanvasTexture(canvas);
        const labelMaterial = new THREE.MeshBasicMaterial({ map: texture });
        const labelGeometry = new THREE.PlaneGeometry(0.3, 0.3);
        const label = new THREE.Mesh(labelGeometry, labelMaterial);
        label.position.set(0, 0.75, 0.31);
        this.mesh.add(label);
        
        this.mesh.position.copy(this.position);
        this.scene.add(this.mesh);
    }

    /**
     * Setup input listeners
     */
    setupInputListeners() {
        if (!this.isLocalPlayer) return;

        window.addEventListener('keydown', (e) => {
            const key = e.key.toUpperCase();
            if (key in this.keysPressed) this.keysPressed[key] = true;
        });

        window.addEventListener('keyup', (e) => {
            const key = e.key.toUpperCase();
            if (key in this.keysPressed) this.keysPressed[key] = false;
        });
    }

    /**
     * Main update loop
     */
    update(deltaTime) {
        if (gameStateManager.getCurrentGameState() === gameStateManager.GameState.Paused) {
            this.currentVelocity.multiplyScalar(0.95);
            return;
        }

        if (this.isLocalPlayer) {
            this.handleInput();
        }

        this.updateStamina(deltaTime);
        this.applyMovement(deltaTime);
        this.handleRotation(deltaTime);
        
        if (this.isDribbling && this.ball) {
            this.applyDribbleControl(deltaTime);
        }

        // Update mesh position
        this.mesh.position.copy(this.position);
        this.mesh.position.y += 0.75; // Adjust for mesh offset
    }

    /**
     * Handle keyboard input
     */
    handleInput() {
        // Get 8-directional input
        let horizontal = 0;
        let vertical = 0;

        if (this.keysPressed.W) vertical += 1;
        if (this.keysPressed.S) vertical -= 1;
        if (this.keysPressed.A) horizontal -= 1;
        if (this.keysPressed.D) horizontal += 1;

        this.inputDirection.set(horizontal, vertical).normalize();

        // Sprint with Shift
        this.isSprinting = this.keysPressed.Shift && this.inputDirection.length() > 0.1 && this.canSprint;

        // Dribbling detection
        if (this.ball) {
            const distanceToBall = this.position.distanceTo(this.ball.getPosition());
            this.isDribbling = 
                distanceToBall < this.dribbleControlRadius && 
                gameStateManager.getCurrentPossession() === this.playerTeam;
        }

        // Passing (Space)
        if (this.keysPressed.Space) {
            if (this.isDribbling) {
                console.log(`Player ${this.playerNumber}: Pass initiated`);
                // Implement pass logic in game manager
            }
        }

        // Shooting (Control)
        if (this.keysPressed.Control) {
            if (this.isDribbling) {
                console.log(`Player ${this.playerNumber}: Shot charging...`);
                // Implement shot logic in game manager
            }
        }
    }

    /**
     * Apply movement physics
     */
    applyMovement(deltaTime) {
        // Determine current max speed
        let currentMaxSpeed = this.isSprinting ? this.maxSprintSpeed : this.maxSpeed;

        // Apply dribble slow-down if holding ball
        if (this.isDribbling) {
            currentMaxSpeed *= this.dribbleSlowDownFactor;
        }

        // Target velocity based on input
        let targetVelocity = new THREE.Vector3(0, 0, 0);
        if (this.inputDirection.length() > 0.1) {
            const moveDirection = new THREE.Vector3(this.inputDirection.x, 0, this.inputDirection.y).normalize();
            targetVelocity = moveDirection.multiplyScalar(currentMaxSpeed);

            // Smooth acceleration
            this.currentVelocity.lerp(targetVelocity, deltaTime * this.acceleration);
        } else {
            // Decelerate when no input
            this.currentVelocity.multiplyScalar(1 - deltaTime * this.deceleration);
        }

        // Apply velocity
        this.position.add(this.currentVelocity.clone().multiplyScalar(deltaTime));
    }

    /**
     * Handle smooth rotation towards movement direction
     */
    handleRotation(deltaTime) {
        if (this.inputDirection.length() < 0.1) return;

        // Calculate target rotation
        const targetRotation = Math.atan2(this.inputDirection.x, this.inputDirection.y);
        
        // Smoothly rotate towards target
        const angleDiff = targetRotation - this.rotation;
        this.rotation += angleDiff * deltaTime * this.rotationSpeed;
        
        // Apply rotation to mesh
        this.mesh.rotation.y = this.rotation;
    }

    /**
     * Apply dribble control to keep ball near player
     */
    applyDribbleControl(deltaTime) {
        const ballPos = this.ball.getPosition();
        const trackingPos = this.position.clone();
        trackingPos.y = ballPos.y; // Adjust for ball height
        
        // Add forward offset based on player velocity
        const forwardOffset = Math.min(this.currentVelocity.length() / this.maxSpeed, 1) * 1;
        const moveDir = this.currentVelocity.clone().normalize();
        if (moveDir.length() > 0) {
            trackingPos.add(moveDir.multiplyScalar(forwardOffset * 0.5));
        }
        
        const distanceVector = trackingPos.clone().sub(ballPos);

        // Pull ball back if drifted
        if (distanceVector.length() > 0.2) {
            const pullForce = distanceVector.normalize().multiplyScalar(this.dribbleDragStrength);
            this.ball.applyForce(pullForce);
        }

        // Lose possession if ball gets too far
        if (distanceVector.length() > this.dribbleMaxBallDistance) {
            this.losePossession();
        }
    }

    /**
     * Update stamina
     */
    updateStamina(deltaTime) {
        if (this.isSprinting && this.currentStamina > 0) {
            this.currentStamina -= this.staminaDrainSprint * deltaTime;
            this.canSprint = this.currentStamina > 10;
        } else {
            this.currentStamina = Math.min(
                this.currentStamina + this.staminaRecovery * deltaTime,
                this.maxStamina
            );
            this.canSprint = true;
        }
    }

    /**
     * Gain ball possession
     */
    gainPossession() {
        this.hasBallPossession = true;
        this.lastTouchTime = gameStateManager.getMatchTimeInSeconds();
        gameStateManager.changePossession(this.playerTeam);
        console.log(`Player ${this.playerNumber} (${this.playerTeam}) gained possession`);
    }

    /**
     * Lose ball possession
     */
    losePossession() {
        this.hasBallPossession = false;
        this.isDribbling = false;
        console.log(`Player ${this.playerNumber} lost possession`);
    }

    /**
     * Execute a ground pass
     */
    performGroundPass(targetDirection, power) {
        if (this.ball && this.isDribbling) {
            this.ball.executeGroundPass(targetDirection.normalize(), power);
            this.losePossession();
        }
    }

    /**
     * Execute a lofted pass
     */
    performLoftedPass(targetDirection, power) {
        if (this.ball && this.isDribbling) {
            this.ball.executeLoftedPass(targetDirection.normalize(), power);
            this.losePossession();
        }
    }

    /**
     * Execute a shot on goal
     */
    performShot(targetDirection, power, accuracyStat = 0.7) {
        if (this.ball && this.isDribbling) {
            this.ball.executeShot(targetDirection.normalize(), power, accuracyStat);
            this.losePossession();
            console.log(`Player ${this.playerNumber} shot with power ${power} and accuracy ${accuracyStat}`);
        }
    }

    // ===== Getters =====
    getStaminaPercent() {
        return this.currentStamina / this.maxStamina;
    }

    getTeam() {
        return this.playerTeam;
    }

    getPlayerNumber() {
        return this.playerNumber;
    }

    getPosition() {
        return this.position.clone();
    }

    getVelocity() {
        return this.currentVelocity.clone();
    }

    getSpeed() {
        return this.currentVelocity.length();
    }

    isDribblingBall() {
        return this.isDribbling;
    }

    hasPossession() {
        return this.hasBallPossession;
    }
}

File: Scripts/PlayerController.cs
using UnityEngine;

/// <summary>
/// PlayerController - Handles all player movement, input, and ball interaction
/// Features:
/// - 8-directional movement with smooth acceleration/deceleration
/// - Momentum-based physics for realistic feel
/// - Dribbling mechanics (ball stays near feet with physics separation)
/// - Sprint capability with stamina drain
/// - Passing and shooting input preparation
/// </summary>
public class PlayerController : MonoBehaviour
{
    // ===== References =====
    private Rigidbody rb;
    private BallPhysics ball;
    private Animator animator;
    [SerializeField] private Transform ballTrackingPoint; // Position where ball should stay near during dribbling

    // ===== Movement Parameters =====
    [SerializeField] private float maxSpeed = 12f;           // Maximum walking speed (m/s)
    [SerializeField] private float maxSprintSpeed = 18f;     // Maximum sprint speed (m/s)
    [SerializeField] private float acceleration = 20f;       // How quickly player reaches max speed
    [SerializeField] private float deceleration = 15f;       // How quickly player slows down
    [SerializeField] private float rotationSpeed = 10f;      // Smooth rotation towards movement direction
    [SerializeField] private float dribbleSlowDownFactor = 0.85f; // Speed reduction when dribbling

    // ===== Dribbling Parameters =====
    [SerializeField] private float dribbleControlRadius = 1.5f;  // How far ball can drift from feet
    [SerializeField] private float dribbleDragStrength = 8f;     // How strongly player pulls ball back
    [SerializeField] private float dribbleMaxBallDistance = 2.5f; // Max distance before losing possession

    // ===== Stamina System =====
    [SerializeField] private float maxStamina = 100f;
    [SerializeField] private float staminaDrainSprint = 30f;   // Stamina drain per second while sprinting
    [SerializeField] private float staminaRecovery = 15f;      // Stamina recovery per second
    private float currentStamina;
    private bool canSprint = true;

    // ===== Input Handling =====
    private Vector2 inputDirection = Vector2.zero;
    private bool isSprinting = false;
    private bool isDribbling = false;
    private Vector3 currentVelocity = Vector3.zero;

    // ===== Player State =====
    [SerializeField] private GameStateManager.Team playerTeam = GameStateManager.Team.TeamA;
    [SerializeField] private bool isLocalPlayer = false;  // Is this player controlled by the player
    [SerializeField] private int playerNumber = 7;        // Jersey number for UI/identification
    private bool hasBallPossession = false;
    private float lastTouchTime = 0f;

    // ===== Initialization =====
    private void Start()
    {
        rb = GetComponent<Rigidbody>();
        animator = GetComponent<Animator>();
        
        // Find ball in scene
        BallPhysics[] balls = FindObjectsOfType<BallPhysics>();
        if (balls.Length > 0)
            ball = balls[0];

        // Create a child object for ball tracking if it doesn't exist
        if (ballTrackingPoint == null)
        {
            GameObject trackingObj = new GameObject("BallTrackingPoint");
            trackingObj.transform.SetParent(transform);
            trackingObj.transform.localPosition = new Vector3(0, 0, 0.5f); // Slightly in front of feet
            ballTrackingPoint = trackingObj.transform;
        }

        currentStamina = maxStamina;
        rb.constraints = RigidbodyConstraints.FreezeRotation; // Prevent physics rotation
    }

    // ===== Update Loop =====
    private void Update()
    {
        if (!isLocalPlayer || GameStateManager.Instance.GetCurrentGameState() == GameStateManager.GameState.Paused)
            return;

        HandleInput();
        UpdateStamina();
        UpdateAnimations();
    }

    // ===== Physics Update =====
    private void FixedUpdate()
    {
        if (!isLocalPlayer || GameStateManager.Instance.GetCurrentGameState() == GameStateManager.GameState.Paused)
        {
            rb.velocity = Vector3.Lerp(rb.velocity, Vector3.zero, Time.fixedDeltaTime * 5f);
            return;
        }

        ApplyMovement();
        HandleRotation();
        
        if (isDribbling && ball != null)
        {
            ApplyDribbleControl();
        }
    }

    // ===== Input Handling =====
    private void HandleInput()
    {
        // Get WASD input (normalized 8-directional movement)
        float horizontal = 0f;
        float vertical = 0f;

        if (Input.GetKey(KeyCode.W)) vertical += 1f;
        if (Input.GetKey(KeyCode.S)) vertical -= 1f;
        if (Input.GetKey(KeyCode.A)) horizontal -= 1f;
        if (Input.GetKey(KeyCode.D)) horizontal += 1f;

        inputDirection = new Vector2(horizontal, vertical).normalized;

        // Sprint with Left Shift
        isSprinting = Input.GetKey(KeyCode.LeftShift) && inputDirection.magnitude > 0.1f && canSprint;

        // Dribbling detection - check if player is close to ball
        if (ball != null)
        {
            float distanceToBall = Vector3.Distance(transform.position, ball.transform.position);
            isDribbling = distanceToBall < dribbleControlRadius && GameStateManager.Instance.GetCurrentPossession() == playerTeam;
        }

        // Passing (Space - sets up pass, released to execute)
        if (Input.GetKeyDown(KeyCode.Space))
        {
            if (isDribbling)
            {
                // Start pass preparation
                Debug.Log($"Player {playerNumber}: Pass initiated");
            }
        }

        // Shooting (Left Ctrl - hold to charge)
        if (Input.GetKeyDown(KeyCode.LeftControl))
        {
            if (isDribbling)
            {
                Debug.Log($"Player {playerNumber}: Shot charging...");
            }
        }
    }

    // ===== Movement Application =====
    private void ApplyMovement()
    {
        // Determine current max speed based on sprint state
        float currentMaxSpeed = isSprinting ? maxSprintSpeed : maxSpeed;

        // Apply dribble slow-down if holding ball
        if (isDribbling)
            currentMaxSpeed *= dribbleSlowDownFactor;

        // Target velocity based on input
        Vector3 targetVelocity = Vector3.zero;
        if (inputDirection.magnitude > 0.1f)
        {
            // Convert input to world space direction
            Vector3 moveDirection = new Vector3(inputDirection.x, 0, inputDirection.y).normalized;
            targetVelocity = moveDirection * currentMaxSpeed;

            // Smooth acceleration towards target speed
            currentVelocity = Vector3.Lerp(currentVelocity, targetVelocity, Time.fixedDeltaTime * acceleration);
        }
        else
        {
            // Decelerate when no input
            currentVelocity = Vector3.Lerp(currentVelocity, Vector3.zero, Time.fixedDeltaTime * deceleration);
        }

        // Apply velocity (maintain Y-axis for gravity)
        rb.velocity = new Vector3(currentVelocity.x, rb.velocity.y, currentVelocity.z);
    }

    // ===== Rotation Handling =====
    private void HandleRotation()
    {
        if (inputDirection.magnitude < 0.1f)
            return; // Don't rotate if no input

        // Calculate target rotation based on input direction
        Vector3 moveDirection = new Vector3(inputDirection.x, 0, inputDirection.y).normalized;
        Quaternion targetRotation = Quaternion.LookRotation(moveDirection, Vector3.up);

        // Smoothly rotate towards target
        transform.rotation = Quaternion.Lerp(transform.rotation, targetRotation, Time.fixedDeltaTime * rotationSpeed);
    }

    // ===== Dribble Control =====
    private void ApplyDribbleControl()
    {
        Vector3 ballPos = ball.transform.position;
        Vector3 trackingPos = ballTrackingPoint.position;
        Vector3 distanceVector = trackingPos - ballPos;

        // Only apply force if ball drifts too far
        if (distanceVector.magnitude > 0.2f)
        {
            Vector3 pullForce = distanceVector.normalized * dribbleDragStrength;
            ball.rb.AddForce(pullForce, ForceMode.Acceleration);
        }

        // Lose possession if ball gets too far
        if (distanceVector.magnitude > dribbleMaxBallDistance)
        {
            LosePossession();
        }

        // Update ball tracking point based on player movement
        // Ball should stay in front of the player when moving
        float forwardOffset = Mathf.Clamp(currentVelocity.magnitude / maxSpeed * 1f, 0.3f, 1f);
        ballTrackingPoint.localPosition = new Vector3(0, 0.1f, forwardOffset);
    }

    // ===== Stamina System =====
    private void UpdateStamina()
    {
        if (isSprinting && currentStamina > 0)
        {
            currentStamina -= staminaDrainSprint * Time.deltaTime;
            canSprint = currentStamina > 10f; // Can't sprint if stamina below 10%
        }
        else
        {
            currentStamina = Mathf.Min(currentStamina + staminaRecovery * Time.deltaTime, maxStamina);
            canSprint = true;
        }
    }

    // ===== Animation Updates =====
    private void UpdateAnimations()
    {
        if (animator == null)
            return;

        // Set movement speed parameter (normalized 0-1)
        float speedPercent = rb.velocity.magnitude / maxSprintSpeed;
        animator.SetFloat("Speed", speedPercent);

        // Set sprint parameter
        animator.SetBool("IsSprinting", isSprinting);

        // Set dribbling parameter
        animator.SetBool("IsDribbling", isDribbling);
    }

    // ===== Ball Interaction =====
    public void GainPossession()
    {
        hasBallPossession = true;
        lastTouchTime = GameStateManager.Instance.GetMatchTimeInSeconds();
        GameStateManager.Instance.ChangePossession(playerTeam);
        Debug.Log($"Player {playerNumber} ({playerTeam}) gained possession");
    }

    public void LosePossession()
    {
        hasBallPossession = false;
        isDribbling = false;
        Debug.Log($"Player {playerNumber} lost possession");
    }

    /// <summary>
    /// Execute a ground pass - faster, more accurate but shorter range
    /// </summary>
    public void PerformGroundPass(Vector3 targetDirection, float power)
    {
        if (ball != null && isDribbling)
        {
            ball.ExecutePass(targetDirection, power, BallPhysics.PassType.Ground);
            LosePossession();
        }
    }

    /// <summary>
    /// Execute a lofted pass - slower, longer range, higher arc
    /// </summary>
    public void PerformLoftedPass(Vector3 targetDirection, float power)
    {
        if (ball != null && isDribbling)
        {
            ball.ExecutePass(targetDirection, power, BallPhysics.PassType.Lofted);
            LosePossession();
        }
    }

    /// <summary>
    /// Execute a shot on goal - high power, adds error based on player stat
    /// </summary>
    public void PerformShot(Vector3 targetDirection, float power, float accuracyStat)
    {
        if (ball != null && isDribbling)
        {
            // Add slight random error based on accuracy (0-1, higher is better)
            float errorRange = (1f - Mathf.Clamp01(accuracyStat)) * 15f; // Up to 15 degrees error
            float randomAngle = Random.Range(-errorRange, errorRange);
            Quaternion errorRotation = Quaternion.AngleAxis(randomAngle, Vector3.up);
            Vector3 adjustedDirection = errorRotation * targetDirection;

            ball.ExecuteShot(adjustedDirection, power);
            LosePossession();
            Debug.Log($"Player {playerNumber} shot with power {power} and accuracy {accuracyStat}");
        }
    }

    // ===== Getters =====
    public float GetStaminaPercent() => currentStamina / maxStamina;
    public bool HasBallPossession() => hasBallPossession;
    public bool IsDribbling() => isDribbling;
    public float GetCurrentSpeed() => rb.velocity.magnitude;
    public Vector3 GetCurrentVelocity() => currentVelocity;
    public GameStateManager.Team GetTeam() => playerTeam;
    public int GetPlayerNumber() => playerNumber;
}
